src/util/shared_exclusive_lock.c

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2006-Present, Redis Ltd.
+ * All rights reserved.
+ *
+ * Licensed under your choice of the Redis Source Available License 2.0
+ * (RSALv2); or (b) the Server Side Public License v1 (SSPLv1); or (c) the
+ * GNU Affero General Public License v3 (AGPLv3).
+*/
+#include "shared_exclusive_lock.h"
+#include <pthread.h>
+#include <stdatomic.h>
+#include <time.h>
+
+#define TIMEOUT_NANOSECONDS 5000 // 5 us in nanoseconds
+
+pthread_mutex_t InternalLock = PTHREAD_MUTEX_INITIALIZER;
+pthread_mutex_t AuxLock = PTHREAD_MUTEX_INITIALIZER;
+pthread_cond_t GILCondition = PTHREAD_COND_INITIALIZER;
+pthread_cond_t AuxLockCondition = PTHREAD_COND_INITIALIZER;
+atomic_bool GILOwned;
+bool InternalLockHeld;
+
+
+void SharedExclusiveLock_Init() {
+  atomic_init(&GILOwned, false);
+  InternalLockHeld = false;
+  pthread_mutex_init(&InternalLock, NULL);
+  pthread_mutex_init(&AuxLock, NULL);
+  pthread_cond_init(&GILCondition, NULL);
+  pthread_cond_init(&AuxLockCondition, NULL);
+}
+
+void SharedExclusiveLock_Destroy() {
+  pthread_mutex_destroy(&InternalLock);
+  pthread_cond_destroy(&GILCondition);
+  pthread_mutex_destroy(&AuxLock);
+  pthread_cond_destroy(&AuxLockCondition);
+}
+
+void SharedExclusiveLock_SetOwned() {
+  atomic_store_explicit(&GILOwned, true, memory_order_release);
+  pthread_cond_broadcast(&GILCondition);
```

Comment:
It's probably better to make sure the signal is sent after the variable was set, so we should use aquire+release (which is the default memory order):
```suggestion
  atomic_store(&GILOwned, true);
  pthread_cond_broadcast(&GILCondition);
```